core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -699,76 +741,135 @@ private void allocateUnassigned() {
-                                         *    1 2 0
-                                         *    2 0 1
-                                         *    0 1 2
-                                         *  such that if we need to tie-break we try to prefer the node holding a shard with the minimal id greater
-                                         *  than the id of the shard we need to assign. This works find when new indices are created since
-                                         *  primaries are added first and we only add one shard set a time in this algorithm.
-                                         */
-                                    if (currentDecision.type() == decision.type()) {
-                                        final int repId = shard.id();
-                                        final int nodeHigh = node.highestPrimary(shard.index().getName());
-                                        final int minNodeHigh = minNode.highestPrimary(shard.getIndexName());
-                                        updateMinNode = ((((nodeHigh > repId && minNodeHigh > repId)
-                                                               || (nodeHigh < repId && minNodeHigh < repId))
-                                                              && (nodeHigh < minNodeHigh))
-                                                             || (nodeHigh > minNodeHigh && nodeHigh > repId && minNodeHigh < repId));
-                                    } else {
-                                        updateMinNode = currentDecision.type() == Type.YES;
-                                    }
-                                } else {
-                                    updateMinNode = true;
-                                }
-                                if (updateMinNode) {
-                                    minNode = node;
-                                    minWeight = currentWeight;
-                                    decision = currentDecision;
-                                }
-                            }
+            if (canShortCircuit(throttledNodes)) {
+                // all nodes are throttled, so we know we won't be able to allocate this round,
+                // so if we are not in explain mode, short circuit
+                return ShardAllocationDecision.no(UnassignedInfo.AllocationStatus.DECIDERS_NO, null);
+            }
+            /* Don't iterate over an identity hashset here the
+             * iteration order is different for each run and makes testing hard */
+            final boolean explain = allocation.debugDecision();
+            Map<String, NodeExplanation> nodeExplanationMap = explain ? new HashMap<>() : null;
+            for (ModelNode node : nodes.values()) {
+                // simulate weight if we would add shard to node
+                float currentWeight = weight.weightShardAdded(this, node, shard.getIndexName());
```

Comment:
now we have to calculate the weight even if we short circuit in the next line due to other reasons. Let's do this calculation lazily even if it requires one more level of `if` nesting.